{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the number of your choice: 43.43\n",
      "Expected integer value. please try again\n"
     ]
    }
   ],
   "source": [
    "i = True\n",
    "while i==True:\n",
    "    try:\n",
    "        n=int(input(\"Enter the number of your choice: \")) \n",
    "        i=False\n",
    "    except:\n",
    "        print(\"Expected integer value. please try again\")\n",
    " \n",
    "max=50\n",
    "diff=max-n\n",
    "\n",
    "if diff > 30:\n",
    "    print(\"Difference is \",diff)\n",
    "    print(\"Square of difference is \",diff**2)\n",
    "else:\n",
    "    print(\"Difference is less than 30!, it is {}\".format(diff))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Introduction to Exception handling - 1. basic block\n",
    "\n",
    "# Try block used to test a block for any kind of awkard inputs\n",
    "\n",
    "# Except block used to handle any kind of awkard inputs\n",
    "\n",
    "# Finally block used to handle a normal execution during any kind of awkard inputs\n",
    "\n",
    "try: # try this code for execution\n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    c= a+b # to add two numbers\n",
    "\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c)) # to print o/p\n",
    "    \n",
    "except: # if exception occurs\n",
    "    print(\"Exceptional input!!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Introduction to Exception handling - 2.  variants\n",
    "\n",
    "# Try block used to test a block for any kind of awkard inputs\n",
    "\n",
    "# Except block used to handle any kind of awkard inputs\n",
    "\n",
    "# Finally block used to handle a normal execution during any kind of awkard inputs\n",
    "\n",
    "try: # try this code for execution\n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    c= a+b # to add two numbers\n",
    "\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c)) # to print o/p\n",
    "    \n",
    "except: # if exception occurs\n",
    "    print(\"Exceptional input!!\")\n",
    "\n",
    "finally: # it will be executed , does not matter if try execpt will generate any kind of exception\n",
    "    print(\"Code executed !!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # enhance the functionality by using exception handling in this program\n",
    "# ask user to input some number\n",
    "# whatever that number is, suppose if user enters 10, we will find the difference of entered number with 50.\n",
    "# if difference is more than 30, we will return the square or double of exact difference of two numbers\n",
    "\n",
    "try:\n",
    "    n=int(input(\"Enter the number of your choice: \"))\n",
    "    max=50\n",
    "    diff=max-n\n",
    "\n",
    "    if diff > 30:\n",
    "        print(\"Difference is \",diff)\n",
    "        print(\"Square of difference is \",diff**2)\n",
    "    else:\n",
    "        print(\"Difference is less than 30!, it is {}\".format(diff))\n",
    "except:\n",
    "    print(\"Something went wrong!!\")\n",
    "    \n",
    "finally:\n",
    "    print(\"Code is executed!!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Introduction to Exception handling - How to raise a error specific message \"Name Error\"\n",
    "\n",
    "# Try block used to test a block for any kind of awkard inputs\n",
    "# Except block used to handle any kind of awkard inputs\n",
    "# Finally block used to handle a normal execution during any kind of awkard inputs\n",
    "\n",
    "try: # try this code for execution\n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    c= a+b # to add two numbers\n",
    "\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c)) # to print o/p\n",
    "\n",
    "except NameError: # It will be raised when Name Error will occur.\n",
    "    print(\"Number input is expected!\")\n",
    "\n",
    "except: # if some other exception occurs, except Name Error.\n",
    "    print(\"Some other exception occurred!!\")\n",
    "\n",
    "finally: # it will be executed , does not matter if try execpt will generate any kind of exception\n",
    "    print(\"Code executed !!\")\n",
    "    \n",
    "# One try block can carry multiple except statements\n",
    "# name specific exception can be mentioned in the code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Introduction to Exception handling - How to raise a error specific message \"\"\n",
    "\n",
    "#Introduction to file handling - 1. How to close a file , check if file is actually closed or not!!\n",
    "\n",
    "try: # to implement exception handling\n",
    "    \n",
    "    f= open(\"MyFolder/myfile.txt\") # file to be opened from different then default location where notebook file is saved.\n",
    "\n",
    "    print(f.read()) # we can use read method for reading the contents of the file upto particlar length\n",
    "\n",
    "    f.close() # to close file object\n",
    "\n",
    "    print(f.read()) # this willl generate an error as we have already closed file object in line no. 7\n",
    "\n",
    "except ValueError: # to take care value Error\n",
    "    print(\"You are trying to access closed file!\")\n",
    "\n",
    "except: # to take care all other kind of errors\n",
    "    print(\"Invalid I/O Operation!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# different kind of errors in python, try to execute line by line\n",
    "\n",
    "#print(5+'5') # TypeError\n",
    "#--------------------------------------------------------------------------------------------\n",
    "#ans=100/0 #ZeroDivisionError\n",
    "#--------------------------------------------------------------------------------------------\n",
    "#val=input(\"Enter val:\"  #KeyboardInterrupt, stop the execution during input statement\n",
    "#--------------------------------------------------------------------------------------------\n",
    "#print \"hello world! # python 2, but its not accepted in Python 3, # SyntaxError\n",
    "#--------------------------------------------------------------------------------------------\n",
    "# list1=[1,5,7,9]\n",
    "# print(list1[4]) #IndexError  as max possible value is 3\n",
    "#--------------------------------------------------------------------------------------------\n",
    "#import randomname #ImportError\n",
    "#--------------------------------------------------------------------------------------------\n",
    "# studdict = {\n",
    "#     \"Name\": \"ABC\"\n",
    "#             }\n",
    "\n",
    "# print(\"Default dictionary\",studdict['Namee']) # KeyError\n",
    "#--------------------------------------------------------------------------------------------\n",
    "# m='Twenty'\n",
    "# print(int(m)) # ValueError\n",
    "#--------------------------------------------------------------------------------------------\n",
    "# m='1'\n",
    "# print(n) # NameError\n",
    "#--------------------------------------------------------------------------------------------\n",
    "# m=1\n",
    "#     if m>1: #IndentationError\n",
    "#         print(\"Random Text\")\n",
    "#     else:\n",
    "#         print(\"1\")\n",
    "#--------------------------------------------------------------------------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Exception handling - on Operators - Arithmetic\n",
    "\n",
    "try:\n",
    "    \n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    # arithmetic operations\n",
    "    c= a+b;    d= a*b;    e= b-a;    f= a/b;    g= a%b;    h= a**b;    i= a//b; # use ; to add multiple statements in one line\n",
    "\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c))\n",
    "    print(\"Multiplication of {} and {} is {}\".format(a,b,d))\n",
    "    print(\"Subtraction of {} and {} is {}\".format(a,b,e))\n",
    "    print(\"Division of {} and {} is {}\".format(a,b,f))\n",
    "    print(\"Modulus of {} and {} is {}\".format(a,b,g))\n",
    "    print(\"Exponentiation of {} and {} is {}\".format(a,b,h))\n",
    "    print(\"Floor division of {} and {} is {}\".format(a,b,i))\n",
    "    \n",
    "except NameError:\n",
    "    print(\"Name Error\")\n",
    "\n",
    "except ZeroDivisionError:\n",
    "    print(\"Zero Division Error\")\n",
    "    \n",
    "except ValueError:\n",
    "    print(\"Value Error\")\n",
    "\n",
    "except SyntaxError:\n",
    "    print(\"Syntax Error\")\n",
    "\n",
    "except:\n",
    "    print(\"Something went wrong!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Introduction to Exception handling - How to display a system defined error message\n",
    "\n",
    "try: \n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    c= a+b\n",
    "\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c)) # to print o/p\n",
    "\n",
    "except NameError as nm: # used to display system defined exception messages to the user\n",
    "    print(nm)\n",
    "\n",
    "except: # if some other exception occurs, except Name Error.\n",
    "    print(\"Some other exception occurred!!\")\n",
    "\n",
    "finally: # it will be executed , does not matter if try execpt will generate any kind of exception\n",
    "    print(\"Code executed !!\")\n",
    "    \n",
    "# One try block can carry multiple except statements\n",
    "# name specific exception can be mentioned in the code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Exception handling - on Operators - Arithmetic to display custom and system defined messages to the user\n",
    "\n",
    "try:\n",
    "    \n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    # arithmetic operations\n",
    "    c= a+b;    d= a*b;    e= b-a;    f= a/b;    g= a%b;    h= a**b;    i= a//b; # use ; to add multiple statements in one line\n",
    "\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c))\n",
    "    print(\"Multiplication of {} and {} is {}\".format(a,b,d))\n",
    "    print(\"Subtraction of {} and {} is {}\".format(a,b,e))\n",
    "    print(\"Division of {} and {} is {}\".format(a,b,f))\n",
    "    print(\"Modulus of {} and {} is {}\".format(a,b,g))\n",
    "    print(\"Exponentiation of {} and {} is {}\".format(a,b,h))\n",
    "    print(\"Floor division of {} and {} is {}\".format(a,b,i))\n",
    "    \n",
    "except NameError as nm:\n",
    "    print(\"Name Error\",nm)\n",
    "\n",
    "except ZeroDivisionError as ze:\n",
    "    print(\"Zero Division Error\",ze)\n",
    "    \n",
    "except ValueError as ve:\n",
    "    print(\"Value Error\",ve)\n",
    "\n",
    "except SyntaxError as se:\n",
    "    print(\"Syntax Error\",se)\n",
    "\n",
    "except:\n",
    "    print(\"Something went wrong!\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Exception handling - on Operators - Arithmetic by jayant\n",
    "# single exception block to display system defined messages to the user\n",
    "\n",
    "try:\n",
    "    \n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    # arithmetic operations\n",
    "    c= a+b;    d= a*b;    e= b-a;    f= a/b;    g= a%b;    h= a**b;    i= a//b; # use ; to add multiple statements in one line\n",
    "\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c))\n",
    "    print(\"Multiplication of {} and {} is {}\".format(a,b,d))\n",
    "    print(\"Subtraction of {} and {} is {}\".format(a,b,e))\n",
    "    print(\"Division of {} and {} is {}\".format(a,b,f))\n",
    "    print(\"Modulus of {} and {} is {}\".format(a,b,g))\n",
    "    print(\"Exponentiation of {} and {} is {}\".format(a,b,h))\n",
    "    print(\"Floor division of {} and {} is {}\".format(a,b,i))\n",
    "    \n",
    "except Exception as mess: # to handle all kin dof exceptions which can occur in the program\n",
    "    print(mess) #to print system defined messages on the screen for the user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Exception handling - on Operators - Arithmetic - how to write try code and normal code separately\n",
    "# single exception block to display system defined messages to the user\n",
    "\n",
    "try:\n",
    "    a= int(input(\"Enter First Number \"))\n",
    "    b= int(input(\"Enter Second Number \"))\n",
    "\n",
    "    # arithmetic operations\n",
    "    c= a+b;    d= a*b;    e= b-a;    f= a/b;    g= a%b;    h= a**b;    i= a//b; # use ; to add multiple statements in one line\n",
    "\n",
    "except Exception as mess: # to handle all kin dof exceptions which can occur in the program\n",
    "    print(mess) #to print system defined messages on the screen for the user\n",
    "\n",
    "else: # it will execute if exception is not occurred in the code\n",
    "    print(\"Sum of {} and {} is {}\".format(a,b,c))\n",
    "    print(\"Multiplication of {} and {} is {}\".format(a,b,d))\n",
    "    print(\"Subtraction of {} and {} is {}\".format(a,b,e))\n",
    "    print(\"Division of {} and {} is {}\".format(a,b,f))\n",
    "    print(\"Modulus of {} and {} is {}\".format(a,b,g))\n",
    "    print(\"Exponentiation of {} and {} is {}\".format(a,b,h))\n",
    "    print(\"Floor division of {} and {} is {}\".format(a,b,i))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
